pydis_site/apps/api/viewsets/bot/nomination.py

```diff
@@ -1,32 +1,224 @@
+    See `GET /bot/nominations/<id:int>`
+
+    #### Status codes
+    - 200: returned on success
+    - 400: returned on failure for the following reasons:
+        - `unnominate_reason` is missing from the request body;
+        - Any other field is present in the request body;
+        - The nomination was already inactiive.
+    - 404: if an infraction with the given `id` could not be found
+    """
     serializer_class = NominationSerializer
-    queryset = Nomination.objects.prefetch_related('author', 'user')
-    frozen_fields = ('author', 'inserted_at', 'user')
+    queryset = Nomination.objects.all()
+    filter_backends = (DjangoFilterBackend, SearchFilter, OrderingFilter)
+    filter_fields = ('user__id', 'actor__id', 'active')
+    frozen_fields = ('id', 'actor', 'inserted_at', 'user', 'unwatched_at', 'active')
+    frozen_on_create = ('unwatched_at', 'unnominate_reason', 'active', 'inserted_at')
 
-    def update(self, request, *args, **kwargs):
+    def create(self, request, *args, **kwargs):
         """
-        DRF method for updating a Nomination.
+        DRF method for creating a Nomination.
 
         Called by the Django Rest Framework in response to the corresponding HTTP request.
         """
+        for field in request.data:
+            if field in self.frozen_on_create:
+                raise ValidationError({field: ['This field cannot be set at creation.']})
+
+        user_id = request.data.get("user")
+        if Nomination.objects.filter(active=True, user__id=user_id):
+            raise ValidationError({'active': ['There can only be one active nomination.']})
 
+        serializer = self.get_serializer(
+            data=ChainMap(
+                request.data,
+                {"active": True}
+            )
+        )
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+
+    def partial_update(self, request, *args, **kwargs):
+        """
+        DRF method for updating a Nomination.
+
+        Called by the Django Rest Framework in response to the corresponding HTTP request.
+        """
         for field in request.data:
             if field in self.frozen_fields:
                 raise ValidationError({field: ['This field cannot be updated.']})
 
         instance = self.get_object()
+
+        if instance.active and request.data.get('unnominate_reason'):
+            raise ValidationError(
+                {'unnominate_reason': ["An active nomination can't have an unnominate reason."]}
+            )
+
+        serializer = self.get_serializer(instance, data=request.data, partial=True)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        return Response(serializer.data)
+
+    @action(detail=True, methods=['patch'])
+    def end_nomination(self, request, pk=None):
```

Comment:
Yeah, it's ugly, I agree. I was debating with myself what a proper name could be, but I think just `end` would do as well, as it would look like: `bot/infractions/1/end`. As an endpoint, that makes sense as well. Same logic goes for `dismiss`. I don't mind either way, but I don't like `unnominate` as a term.

Reply:
As much as I like short words for convenience, I think `dismiss` makes more sense than `end`.

Reply:
Hmm, for me, `end` makes more sense than `dismiss`, but I don't really care either way.

Reply:
I've changed it to `dismiss` in the latest push.

Reply:
this is really nitpicky but I think you `withdraw` a nomination, you don't dismiss it.

Reply:
`withdraw` in the context of nominations is always negative right? For us, we end the nomination period both when the nominee is accepted and rejected for helper. We're kinda nitpicky here, but better to have a correct term now than later. It's fairly easy to change, so I don't care what we settle on.

Reply:
That is a good point. In which case, `dismiss` is negative too. `End` doesn't make sense because a nomination isn't something that "starts", at least in the sense that one would not say "I started a nomination". To me, `withdraw` only makes sense if it's coming from the person that originally nominated that someone. For everyone else, it's `dismiss`:

> Jane nominated Bob but withdrew it two days later.
> The board would have dismissed the nomination for Bob anyway because he's stinky.

That being said, I remembered another word: `rescind`. To me, this sounds OK coming from both, though it still has a negative connotation.

Reply:
I don't have a problem with end. It implies an end to the nomination process, sounds completely neutral. I agree that both withdraw and dismiss (and rescind) have too much of a negative connotation to be applicable.

Reply:
Though it isn't perfect in my eyes (and it's not very important that it should be in the first place), I cannot come up with anything better. `End` seems like the best choice right now.

Reply:
I'll change this back to `end` then.